### DISCUSSION TITLE
dev env vs CI env

### DISCUSSION THREAD

**Discussion by tomdavidson** [numtide/devshell#236] [General] [answered]
nix newb here, this might be obvious to others

Im diggin devshell and just barey starting to tinker with it. I dont need devshell, githooks, and whatnot  in CI but I do need things like dprint cargo-make, pnpm, node, etc.

I think I need a devshell flake that also uses a CI flake? Or can I have I but if CI=true then I skip the devShell (would this be a conditional output?)?

Can I have pure but still have pkgs defined outside the devShell scope?

Whats is a right way to solve this?

**Comment by zimbatm** [comment, marked as answer]
Good question!

One way is to split the devshell in two:
```nix
{
  inputs.devshell.url = "github:numtide/devshell";
  inputs.flake-utils.url = "github:numtide/flake-utils";

  outputs = { self, flake-utils, devshell, nixpkgs }:
    flake-utils.lib.eachDefaultSystem (system:
      let
        pkgs = import nixpkgs {
          inherit system;

          overlays = [ devshell.overlay ];
        };
      in
     {
      devShells.default = pkgs.devshell.mkShell {
        imports = [ (pkgs.devshell.importTOML ./devshell.toml) ];
      };
      devShells.ci = pkgs.devshell.mkShell {
        imports = [ (pkgs.devshell.importTOML ./devshell-ci.toml) ];
      }
    });
}
```
Then use `nix develop` for development and `nix develop .#ci` for the CI environment.

You could also have a `(pkgs.devshell.importTOML ./devshell.shared.toml)` in both imports to avoid repeating some of the dependencies.

Hope that helps!

**Comment by tomdavidson** [comment]
Oh that looks way easier than expected :D - thank you @zimbatm  The language still throws me but Ill get sorted - eventually. 

A shell for CI seems intuitive and more "normal".  Should nix-shell be used for CI instead of devshell? A flake app that preforms the needed CI tasks rather than use a shell?

Subsequent `nix develop .#ci  --command ` invocations in CI will each create a new shell, restore cache, run the provided command, update cache, destroy the created shell? Or is it better to enter the shell ?

If I use same package in each shell do they share cache (only matters if using remote cache)? Do they have to declared outside of an individual shell?

**Reply by zimbatm** [reply]
Another valid way is to expose some of the commands as flake apps and then use `nix run .#mytaskname`. This has the advantage of loading the strict minimum of dependencies for the given task. The downside is that it doesn't have the devshell environment variables, which might be needed for the script to run.
